IPv6 link local support

cspiel1 opened the pull request:
In order to use an IPv6 link local address as destination address
for UDP/TCP packets, the sin6_scope_id needs to be set to the used network
interface.

This PR adds the support for IPv6 link local addresses for sip and http.

SIP calls to IPv6 link local addresses and answering calls from IPv6 link local remote will be possible.

sreimers commented on src/net/net.c:
```diff
@@ -95,6 +95,48 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int net_check_conn(const struct sa *src, const struct sa *dst)
```
Can you add a test case for `net_check_conn`? Something similar to: https://github.com/baresip/retest/commit/b9d03176436834c79e93774336526c7eb7d317c5

sreimers commented on src/net/net.c:
```diff
@@ -95,6 +95,48 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
+		if (!err)
+			break;
+
+		sa_set_port(&srctmp, port);
```
Shouldn`t `sa_set_port` before udp_listen? Or is listening on src original port in every case harmless?

sreimers replied:
Is the loop needed? Maybe listening on port 0 to get a free port from system is enough?

sreimers commented on src/net/net.c:
```diff
@@ -95,6 +95,48 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
+	if (err)
+		return err;
+
+	err = udp_connect(us, &dsttmp);
```
![Bildschirmfoto von 2021-06-08 13-00-22](https://user-images.githubusercontent.com/296762/121175010-e0428600-c85a-11eb-8625-435dbb04a9bd.png)

Its very unlikely but `make clang` complains about this.

cspiel1 replied:
Found the `net_check_conn()` does not check anything useful. We can use `net_dst_source_addr_get()` instead, which retrieves a source address for which a given destination is reachable from.

I remove the commits that adds `net_check_conn()`.

sreimers commented on src/net/net.c:
```diff
@@ -95,6 +95,41 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
+	err = udp_connect(us, &dsttmp);
+	if (err)
+		goto out;
+out:
```
Since there is no other goto out, we can remove it completly:

```c
err = udp_connect(us, &dsttmp);

mem_deref(us);
return err;
```

cspiel1 commented:
Should we add a unit test that collects local IP addresses?

alfredh commented on include/re_sa.h:
```diff
@@ -59,5 +59,12 @@ bool     sa_is_linklocal(const struct sa *sa);
+int      sa_cpy_scopeid(struct sa *sa_dst, const struct sa *sa_src);
+bool     sa_is_ipv6ll(const struct sa *sa);
+#endif
+
```
we should not use ifdefs in the public api, because it makes the ABI different depending on the ifdef.

on a side note, we should consider remove ifdef HAVE_INET6 and always have it compiled in.
ipv6 is today pretty much standard everywhere.

cspiel1 replied:
Will remove this for the API functions. Completely removing the switch should be another PR.

alfredh commented on include/re_sa.h:
```diff
@@ -59,5 +59,12 @@ bool     sa_is_linklocal(const struct sa *sa);
 bool     sa_is_any(const struct sa *sa);
 
+#ifdef HAVE_INET6
+void     sa_set_scopeid(struct sa *sa, int scopeid);
```
the scope_id is of type `uint32_t` in the posix api...

should we use the same ?

cspiel1 replied:
Ah, yes!

alfredh commented on include/re_sa.h:
```diff
@@ -59,5 +59,12 @@ bool     sa_is_linklocal(const struct sa *sa);
+#ifdef HAVE_INET6
+void     sa_set_scopeid(struct sa *sa, int scopeid);
+int      sa_scopeid(const struct sa *sa);
+int      sa_cpy_scopeid(struct sa *sa_dst, const struct sa *sa_src);
```
I think this should be baked into `sa_cpy`

cspiel1 replied:
It is already done via memcpy.
`sa_cpy_scopeid()` is needed to copy the scope ID of a local IPv6ll address to a destination IPv6ll address. As far as I know this selects the correct network interface. This has also be done if there is only one network interface where the scope ID is often 2 (not zero).

alfredh commented on src/sa/sa.c:
```diff
@@ -591,3 +591,42 @@ bool sa_is_any(const struct sa *sa)
+	if (!sa)
+		return;
+
+	sa->u.in6.sin6_scope_id = scopeid;
```
the AF type must be checked here.

if the AF is not INET6, it will access the wrong element in the union

alfredh commented on src/sa/sa.c:
```diff
@@ -591,3 +591,42 @@ bool sa_is_any(const struct sa *sa)
+	if (!sa)
+		return 0;
+
+	return sa->u.in6.sin6_scope_id;
```
also check AF here

alfredh commented on src/sa/sa.c:
```diff
@@ -591,3 +591,42 @@ bool sa_is_any(const struct sa *sa)
+}
+
+
+bool sa_is_ipv6ll(const struct sa *sa)
```
not sure if this function is needed.

the check can be done in the application

cspiel1 replied:
Since it is used several times it reduces the code. Your next question could make it obsolete.

alfredh commented on src/sip/request.c:
```diff
@@ -207,6 +211,30 @@ static int request(struct sip_request *req, enum sip_transp tp,
 }
 
 
+static int dst_set_scopeid(struct sa *dst, struct sip_request *req)
```
I am not sure if the application should have knowledge about scope id.

this should be done on a low-level like in sa or net

cspiel1 replied:
The scope ID has to be copied from chosen `laddr` to the destination address. @juha-h and me have already thought about choosing the `laddr` in a very low level. But it has to be put into the SDP.  So in order to make IPv6ll work:
- We have a destination IPv6ll address.
- Loop over all local IPv6ll addresses.
- Copy scope ID from local address to destination address.
- Test if there is a route from local to destination address.
- The found `laddr` has to be put into the SDP.
- Then the destination address has the correct scope ID.

alfredh commented:
for the sa api, I think only two functions are needed:

```c
int sa_set_scopeid(struct sa *sa, uint32_t scope_id);
uint32_t sa_scopeid(const struct sa *sa);
```

the `struct sockaddr_in6` type is public, so I am not sure they are needed at all...

cspiel1 commented:
> for the sa api, I think only two functions are needed:
> 
> ```c
> int sa_set_scopeid(struct sa *sa, uint32_t scope_id);
> uint32_t sa_scopeid(const struct sa *sa);
> ```
> 
> the `struct sockaddr_in6` type is public, so I am not sure they are needed at all...

I would prefer the `sa_cpy_scopeid`. The getter is only to make debugging easier. It could be dropped. But if you prefer I could write:
```
sa_set_scopeid(dst, sa_scopeid(src));
```
instead of
```
sa_cpy_scopeid(dst, src);
```

alfredh commented:
the patch is a bit comprehensive, so I suggest to split up the patch into logical parts.

I would like to suggest a separate PR for the sa changes:

```c
void     sa_set_scopeid(struct sa *sa, uint32_t scopeid);
uint32_t sa_scopeid(const struct sa *sa);
```

alfredh commented on src/net/net.c:
```diff
@@ -83,7 +84,11 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
-	err = udp_connect(us, dst);
+	dsttmp = *dst;
+	if (!sa_port(&dsttmp))
+		sa_set_port(&dsttmp, 53);
```
I dont think we should hardcode port 53 (DNS) into the library

sreimers replied:
This is only a dummy connection, so no real traffic. Its only for setting up the socket. macOS refuses port 0 as destination port.

alfredh replied:
what if the routing table has different rules for port 53 and port 5060 ?

it would make sense to use the same port number as the application would use.

sreimers replied:
Good point, but its only a fallback if the application does not provide a valid dst port. So maybe returning a error code for dst port 0 is a better option and let the application handle this case.

cspiel1 replied:
ok. Then I drop this and will handle this in retest. And we have to solve this also in ua.c.

cspiel1 commented:
> the patch is a bit comprehensive, so I suggest to split up the patch into logical parts.
> 
> I would like to suggest a separate PR for the sa changes:
> 
> ```c
> void     sa_set_scopeid(struct sa *sa, uint32_t scopeid);
> uint32_t sa_scopeid(const struct sa *sa);
> ```

https://github.com/baresip/re/pull/108

alfredh commented:
thanks!

can you please also rebase this patch on top of master ?

alfredh commented on include/re_net.h:
```diff
@@ -40,7 +40,8 @@ struct sa;
 /* Net generic */
 int  net_hostaddr(int af, struct sa *ip);
-int  net_dst_source_addr_get(struct sa *dst, struct sa *ip);
+int  net_dst_source_addr_get(const struct sa *dst, struct sa *ip);
```
can you please make this change into a separate PR ?

is it just changing the type from non-const to const ?

cspiel1 replied:
Now only making parameter `dst` to `const`. PR: https://github.com/baresip/re/pull/109

alfredh commented:
I would also like to suggest that we split the patch into logical parts;

- linklocal for SIP
- linklocal for HTTP

smaller patches are easier to review and test.
if we find an error later, it is also easier to revert.

cspiel1 commented:
Removed the HTTP changes.

alfredh commented on src/net/net.c:
```diff
@@ -95,6 +95,22 @@ int net_dst_source_addr_get(struct sa *dst, struct sa *ip)
+		return ECONNREFUSED;
+
+	return 0;
+}
```
I dont think that this function belongs in a library.

it is just comparing one IP with another IP. this can be done in the application

alfredh commented on src/sip/reply.c:
```diff
@@ -16,6 +16,37 @@
+	if (err) {
+		DEBUG_WARNING("could not get laddr for %j\n", dst);
+		return err;
+	}
```
will this work if you have two or more interfaces with IPv6-LL ?

example:
```
	inet6 fe80::40f7:f4ff:fe8f:20f0%awdl0 prefixlen 64 scopeid 0xa 
	inet6 fe80::f7e:ea28:f0b7:2bf9%utun0 prefixlen 64 scopeid 0xb 
```

`sip_transp_laddr` will return the first match.

if you are sending a SIP packet to an IPv6 device that is only reachable on network interface
number 2, how will this work ?

cspiel1 replied:
In `transp_find()` we have:
```
		if (net_dst_source_addr_get(&dsttmp, &src))
			continue;

		if (!sa_cmp(&src, laddr, SA_ADDR))
			continue;
```
this will filter away the first IPv6ll address. It tests if there is a route from `laddr` to `dsttmp`. The second IPv6ll is returned.

cspiel1 commented:
As result to @alfredh-s [comment in the baresip PR](https://github.com/baresip/baresip/pull/1473#issuecomment-862245043)

- Squashed the whole PR into two simple commits. Should I make separate PRs also for libre?
- Could set the IPv6 scope ID directly before the send operation in transp.c. Removed all other calls to `sa_set_scopeid()`.

cspiel1 commented:
Don't know how to fix this build problem.
```
Err:21 https://packages.microsoft.com/ubuntu/20.04/prod focal Release
25
  Could not connect to packages.microsoft.com:443 (40.74.238.15), connection timed out [IP: 40.74.238.15 443]
```

sreimers commented:
I triggered re-build

alfredh commented on src/sip/transp.c:
```diff
@@ -145,16 +151,33 @@ static const struct sip_transport *transp_find(struct sip *sip,
+		sa_set_scopeid(&dsttmp, sa_scopeid(laddr));
+
+		if (net_dst_source_addr_get(&dsttmp, &src))
+			continue;
```
is it so that the route check is only done for link-local ?

if so the baresip PR cannot really be applied, as it assumes that multiple addr
for all address types work.

cspiel1 replied:
no, the route check is done for each `laddr`.

cspiel1 commented:
@sreimers I fixed the test in retest branch ipv6ll_support. Please could you re-run the checks?!

sreimers commented:
Looks like it still failing under macOS, I'll take a closer look later.

cspiel1 commented:
On OSX `getaddrinfo()` makes from
```
fe80:0001:0000:0000:0215:58ff:fe2d:90ab
```
this address
```
fe80:0000:0000:0000:0215:58ff:fe2d:90ab
```
The fist might also be invalid.

cspiel1 commented:
I replaced the first commit which added an invalid IPv6ll address.

sreimers commented:
> Is it possible that a successful retest build+checks initiates the checks of an associated re PR?

I re-build manually, looks good now. I will have a closer look on the github action event API, maybe there is a simple solution.

cspiel1 commented:
thanks!

cspiel1 commented:
@alfredh Should we start with this PR in libre? It is already small and I added also tests as retest PR which are all green.
I suggest to review and keep the two commits like they are.

Should I put the two commits into separate PRs?

alfredh commented:
I suggest to merge it, if the code is working ..

cspiel1 commented:
Thanks, I rebase again and start the tests. If it becomes green, everything is ready for the merge.
